### ISSUE TITLE
Unexpected HtmlNode.OuterHtml behaviour when adding one node as a child of another

### ISSUE THREAD

**Issue by alexalok** [zzzprojects/html-agility-pack#250]
In the following example, **text** variable has a value of _"\<outerNode/\>"_, while _"\<outerNode\>\<innerNode/\>\<outerNode/\>"_ is expected. Perhaps this is an expected behavior and I'm doing it wrong?

```c#
var internalNode = HtmlNode.CreateNode("<innerNode/>");
var outerNode = HtmlNode.CreateNode("<outerNode/>");
outerNode.ChildNodes.Add(internalNode);
string text = outerNode.OuterHtml; //<outerNode/>
```

**UPD**

_outerNode_ clone will behave as expected, as demonstrated below:

```C#
var outerNodeClone = outerNode.Clone();
string clonedNodeText = outerNodeClone.OuterHtml; //<outernode><innernode></innernode></outernode>
```

**Comment by JonathanMagnan** [comment]
Hello @alexalok ,

Thank you for reporting, we will look at it.

It seems there is some missing `SetChanged` call when adding or removing the collection. We can usually fix this issue by adding the call but we want to investigate first about side impact of this change.

Best Regards,

Jon

**Comment by JonathanMagnan** [comment]
Hello @alexalok ,

My developer explained me the current behavior.

The code expect you use `AppendChild` instead of `ChildNodes.Add(internalNode)` to add the node correctly.

While it's possible to add it directly with `Add`, that method doesn't call the `SetChanged`.

Could you try to use `AppendChild` instead which is the right way to add a child node.

```csharp
var internalNode = HtmlNode.CreateNode("<innerNode/>");
var outerNode = HtmlNode.CreateNode("<outerNode/>");
outerNode.AppendChild(internalNode);//ChildNodes.Add(internalNode);
string text = outerNode.OuterHtml; //<outerNode/>

var outerNodeClone = outerNode.Clone();//CloneNode(true);
string clonedNodeText = outerNodeClone.OuterHtml; 
```

Let me know if that's working.

Best Regards,

Jonathan

**Comment by alexalok** [comment]
Thanks, @JonathanMagnan, that does resolve my issue. 

if `AppendChild` is a preferrable way of adding child nodes, what is the use-case of `ChildNodes.Add`? If it is to be used only internally, why is it public?

**Comment by JonathanMagnan** [comment]
Hello @alexalok ,

I, unfortunately, don't know this answer (I'm not the original owner of this library). 

Some people may have asked to make it public for their use.

For being backward-compatible, it's a thing that we cannot change (to make it internal).

**Comment by alexalok** [comment]
Okay, thanks a lot for your answers.

**Comment by alexalok** [comment]
@JonathanMagnan, I have noticed another strange behavior which is similar to the one I previously reported so I'll reopen this issue instead of creating a new one.

Please have a look at the following snippet:

```C#
var internalNode = HtmlNode.CreateNode("<innerNode/>");
var outerNode    = HtmlNode.CreateNode("<outerNode/>");
outerNode.AppendChild(internalNode);

var clonedOuterNode = outerNode.Clone();
Debug.Assert(clonedOuterNode.FirstChild.OuterLength == clonedOuterNode.FirstChild.OuterHtml.Length); //false
```

Basically, if you clone a node with a child and then access cloned node's first child `OuterLength` property, it always returns `0`.

**Comment by JonathanMagnan** [comment]
Thank you, we will look at it.

**Comment by JonathanMagnan** [comment]
Hello @alexalok ,

We made some investigation but we finally gave up since we created too much side effect.

The length is currently only set when you load the `HTML`. The library has never been build to use the length with create/clone.

When you create/clone a node, those attributes are currently not copied/created.

After looking at it, we choose to leave it like this.

Let me know if that explains correctly the cause of the issue.

Best Regards,

Jonathan
